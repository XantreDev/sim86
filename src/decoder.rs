use std::u8;

use super::structs::*;

fn decode_reg(reg: u8, w: u8) -> &'static Register {
    #[rustfmt::skip]
        const W_TO_REG: &'static[Register] = &[
            Register::Al, Register::Cl, Register::Dl, Register::Bl, Register::Ah, Register::Ch , Register::Dh, Register::Bh, // w = 0
            Register::Ax, Register::Cx, Register::Dx, Register::Bx, Register::Sp, Register::Bp, Register::Si, Register::Di, // w = 1
        ];

    &W_TO_REG[usize::from(w) * 8 + usize::from(reg)]
}

fn decode_eac<T: Iterator<Item = u8>>(content_iter: &mut T, _mod: u8, rm: u8) -> Operand {
    let is_direct_address = _mod == 0b00 && rm == 0b110;
    let displacement_bytes = if is_direct_address { 2 } else { _mod };

    let displacement_number: Option<i16> = match displacement_bytes {
        2 => Some(
            (u16::from(content_iter.next().unwrap()) | u16::from(content_iter.next().unwrap()) << 8)
                as i16,
        ),
        1 => Some(i16::from((content_iter.next().unwrap()) as i8)),
        _ => None,
    };

    if is_direct_address {
        let Some(displacement_number) = displacement_number else {
            panic!("required direct address");
        };
        return Operand::EAC(None, None, Some(displacement_number));
    }

    static FIRST_ADDRESS: &'static [Register] = &[
        Register::Bx,
        Register::Bx,
        Register::Bp,
        Register::Bp,
        Register::Si,
        Register::Di,
        Register::Bp,
        Register::Bx,
    ];
    static SECOND_ADDRESS: &'static [Option<Register>] = &[
        Some(Register::Si),
        Some(Register::Di),
        Some(Register::Si),
        Some(Register::Di),
        None,
        None,
        None,
        None,
    ];
    // const FIRST_ADDRESS: &'static [Register] = &["bx+si", "bx+di", "bp+si", "bp+di", "si", "di", "bp", "bx"} // rm2registers
    let a_reg = &FIRST_ADDRESS[usize::from(rm)];
    let b_reg = &SECOND_ADDRESS[usize::from(rm)];

    Operand::EAC(Some(*a_reg), *b_reg, displacement_number)
}

fn rm_to_reg<T: Iterator<Item = u8>>(
    op_code: OpCode,
    content_iter: &mut T,
    first: &u8,
    second: &u8,
) -> Instruction {
    let w = first & 0b1;
    let d = (first >> 1) & 0b1;

    let reg = (second >> 3) & 0b111;
    let rm = second & 0b111;
    let _mod = second >> 6 & 0b11;

    let address_or_reg = match _mod {
        0b00..=0b10 => decode_eac(content_iter, _mod, rm),
        0b11 => Operand::Register(*decode_reg(rm, w)),
        _ => {
            panic!("invalid mod processing");
        }
    };

    let reg_name = Operand::Register(*decode_reg(reg, w));

    Instruction {
        left_operand: if d == 0b1 { reg_name } else { address_or_reg },
        right_operand: if d == 0b1 { address_or_reg } else { reg_name },
        flags: if w == 0b1 {
            InstructionFlags::Wide
        } else {
            InstructionFlags::empty()
        },
        op_code,
    }
}

fn immediate_to_rm<T: Iterator<Item = u8>>(
    op_code: OpCode,
    first: &u8,
    second: &u8,
    content_iter: &mut T,
    sign_extend: bool,
) -> Instruction {
    // let sign_extend = explicit_sign_extend.unwrap_or(false);
    let wide = first & 0b1;

    let _mod = second >> 6 & 0b11;
    let rm = second & 0b111;

    let address = match _mod {
        0b00..=0b10 => decode_eac(content_iter, _mod, rm),
        0b11 => Operand::Register(*decode_reg(rm, wide)),
        _ => {
            panic!("invalid mod processing");
        }
    };

    let immediate = {
        let mut init = 0;
        init |= u16::from(content_iter.next().unwrap());

        if wide == 0b1 && sign_extend && (init >> 7) & 0b1 == 1 {
            init |= (u8::MAX as u16) << 8;
        } else if wide == 0b1 && !sign_extend {
            init |= u16::from(content_iter.next().unwrap()) << 8;
        }

        Operand::Immediate(init as i16)
    };

    Instruction {
        op_code,
        flags: if wide == 0b1 {
            InstructionFlags::Wide
        } else {
            InstructionFlags::empty()
        },
        left_operand: address,
        right_operand: immediate,
    }
}

fn mov_immediate_to_reg<T: Iterator<Item = u8>>(
    content_iter: &mut T,
    first: &u8,
    second: &u8,
) -> Instruction {
    assert!(first >> 4 == 0b1011);

    let w = first >> 3 & 0b1;
    let reg = first & 0b111;

    let immediate: i16 = if w == 1 {
        (u16::from(*(second)) | (u16::from(content_iter.next().unwrap()) << 8)) as i16
    } else {
        i16::from(*(second) as i8)
    };

    let reg_name = decode_reg(reg, w);

    // format!("mov {}, {}", reg_name, data)
    Instruction {
        op_code: OpCode::Mov,
        left_operand: Operand::Register(*reg_name),
        right_operand: Operand::Immediate(immediate),
        flags: if w == 0b1 {
            InstructionFlags::Wide
        } else {
            InstructionFlags::empty()
        },
    }
}

fn mov_acc<T: Iterator<Item = u8>>(content_iter: &mut T, first: &u8, second: &u8) -> Instruction {
    assert!(first >> 2 == 0b101000);
    let w = first & 0b1;
    let to_acc = (first >> 1 & 0b1) == 0;

    let reference_address: u16 = if w == 1 {
        u16::from(*(second)) | (u16::from(content_iter.next().unwrap()) << 8)
    } else {
        *(second) as u16
    };

    let flags = if w == 0b1 {
        InstructionFlags::Wide
    } else {
        InstructionFlags::empty()
    };
    if to_acc {
        // format!("mov ax, [{}]", displacement_number)
        Instruction {
            op_code: OpCode::Mov,
            left_operand: Operand::Register(Register::Ax),
            flags,
            right_operand: Operand::Reference(reference_address),
        }
    } else {
        // mov [{}], ax
        Instruction {
            op_code: OpCode::Mov,
            left_operand: Operand::Reference(reference_address),
            flags,
            right_operand: Operand::Register(Register::Ax),
        }
    }
}

fn im_to_acc<T: Iterator<Item = u8>>(
    op_code: OpCode,
    content_iter: &mut T,
    first: &u8,
    second: &u8,
) -> Instruction {
    let w = first & 0b1;

    let immediate: i16 = if w == 1 {
        (u16::from(*(second)) | (u16::from(content_iter.next().unwrap()) << 8)) as i16
    } else {
        i16::from(*(second) as i8)
    };

    Instruction {
        left_operand: Operand::Register(if w == 0 { Register::Al } else { Register::Ax }),
        right_operand: Operand::Immediate(immediate),
        flags: if w == 1 {
            InstructionFlags::Wide
        } else {
            InstructionFlags::empty()
        },
        op_code,
    }
    // format!("al, {}", displacement_number)
}

fn parse_jump(op_code: OpCode, second: &u8) -> Instruction {
    let signed_second = *second as i8;

    Instruction {
        flags: InstructionFlags::empty(),
        op_code,
        left_operand: Operand::JumpDisplacement(signed_second),
        right_operand: Operand::Empty,
    }
}

pub fn decode<T: Iterator<Item = u8>>(
    first: &u8,
    second: &u8,
    content_iter: &mut T,
) -> Option<Instruction> {
    let get_arithmetic_op_code = |byte: &u8| match (byte >> 3) & 0b111 {
        0b000 => OpCode::Add,
        0b101 => OpCode::Sub,
        0b111 => OpCode::Cmp,
        _ => panic!("unknown arhithmetic operand {}", byte >> 3 & 0b111),
    };
    match first {
        // movs
        0b1010_0000..=0b1010_0011 => Some(mov_acc(content_iter, first, second)),
        0b1011_0000..=0b1011_1111 => Some(mov_immediate_to_reg(content_iter, first, second)),
        0b1000_1000..=0b1000_1011 => Some(rm_to_reg(OpCode::Mov, content_iter, first, second)),
        0b1100_0110..=0b1100_0111 => Some(immediate_to_rm(
            OpCode::Mov,
            first,
            second,
            content_iter,
            false,
        )),

        0b100000_00..=0b100000_11 => Some(immediate_to_rm(
            get_arithmetic_op_code(second),
            first,
            second,
            content_iter,
            (first >> 1) & 0b1 == 0b1,
        )),

        0b00_000_0_00..=0b00_000_0_11
        | 0b00_101_0_00..=0b00_101_0_11
        | 0b00_111_0_00..=0b00_111_0_11 => Some(rm_to_reg(
            get_arithmetic_op_code(first),
            content_iter,
            first,
            second,
        )),
        0b00_000_10_0..=0b00_000_10_1
        | 0b00_101_10_0..=0b00_101_10_1
        | 0b00_111_10_0..=0b00_111_10_1 => Some(im_to_acc(
            get_arithmetic_op_code(first),
            content_iter,
            first,
            second,
        )),

        // jumps
        0b0111_0100 => Some(parse_jump(OpCode::Je, second)),
        0b0111_1100 => Some(parse_jump(OpCode::Jl, second)),
        0b0111_1110 => Some(parse_jump(OpCode::Jle, second)),
        0b0111_0010 => Some(parse_jump(OpCode::Jb, second)),
        0b0111_0110 => Some(parse_jump(OpCode::Jbe, second)),
        0b0111_1010 => Some(parse_jump(OpCode::Jp, second)),
        0b0111_0000 => Some(parse_jump(OpCode::Jo, second)),
        0b0111_1000 => Some(parse_jump(OpCode::Js, second)),
        0b0111_0101 => Some(parse_jump(OpCode::Jne, second)),
        0b0111_1101 => Some(parse_jump(OpCode::Jnl, second)),
        0b0111_1111 => Some(parse_jump(OpCode::Jnle, second)),
        0b0111_0011 => Some(parse_jump(OpCode::Jnb, second)),
        0b0111_0111 => Some(parse_jump(OpCode::Jnbe, second)),
        0b0111_1011 => Some(parse_jump(OpCode::Jnp, second)),
        0b0111_0001 => Some(parse_jump(OpCode::Jno, second)),
        0b0111_1001 => Some(parse_jump(OpCode::Jns, second)),
        0b1110_0010 => Some(parse_jump(OpCode::Loop, second)),
        0b1110_0001 => Some(parse_jump(OpCode::Loopz, second)),
        0b1110_0000 => Some(parse_jump(OpCode::Loopnz, second)),
        0b1110_0011 => Some(parse_jump(OpCode::Jcxz, second)),
        _ => None,
    }
}

pub fn decode_instruction<T: Iterator<Item = u8>>(content_iter: &mut T) -> Option<Instruction> {
    let Some(first) = content_iter.next() else {
        return None;
    };
    let second = content_iter.next().unwrap();
    let Some(instruction) = decode(&first, &second, content_iter) else {
        panic!("failed to decode {} {}", first, second)
    };
    Some(instruction)
}
