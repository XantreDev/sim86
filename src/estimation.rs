use super::structs::*;

pub enum Architecture {
    X8086,
    X8088,
}

fn estimate_eac(operand: &Operand) -> u8 {
    let Operand::EAC(reg1, reg2, displ) = operand else {
        panic!("unexpected {:?}", operand)
    };

    // segment overwrites is not supported
    // MOV AX, ES:[BX+SI]
    match (reg1, reg2, displ) {
        (None, None, Some(_)) => 6,
        (
            Some(
                // base or index
                Register::Bx
                | Register::Bl
                | Register::Bh
                | Register::Bp
                | Register::Si
                | Register::Di,
            ),
            None,
            None,
        ) => 5,
        (
            Some(
                // base or index
                Register::Bx
                // is it even possible to have parital registers here?
                | Register::Bl
                | Register::Bh
                | Register::Bp
                | Register::Si
                | Register::Di,
            ),
            None,
            Some(_),
        ) => 9,
        (
            Some(
                Register::Bp
            ),
            Some(
                Register::Di
            ),
            None
        ) | (
            Some(
                Register::Bx
            ),
            Some(
                Register::Si
            ),
            None
        ) => 7,
        (
            Some(
                Register::Bp
            ),
            Some(
                Register::Si
            ),
            None
        ) | (
            Some(
                Register::Bx
            ),
            Some(
                Register::Di
            ),
            None
        ) => 8,
        (
            Some(
                Register::Bp
            ),
            Some(
                Register::Di
            ),
            Some(_)
        ) | (
            Some(
                Register::Bx
            ),
            Some(Register::Si),
            Some(_)
        ) => 11,
        (
            Some(Register::Bp),
            Some(Register::Si),
            Some(_)
        ) | (
            Some(Register::Bx),
            Some(Register::Di),
            Some(_)
        ) => 12,
        _ => panic!("invariant EAC")
    }
}

enum TransferAddress {
    Word(u16),
    Byte(u16),
}

impl CycleEstimation {
    #[inline(always)]
    pub fn estimate_trasfer_tax(
        &self,
        address: TransferAddress,
        architecture: &Architecture,
    ) -> u32 {
        let x8088_odd_operand_transfer_cost = 4u32;
        let cost = match (architecture, address) {
            (Architecture::X8088, TransferAddress::Word(address)) if address % 2 != 0 => {
                x8088_odd_operand_transfer_cost
            }
            _ => 0,
        };

        cost * self.transfers as u32
    }
}

struct CycleEstimation {
    base: u8,
    transfers: u8,
    eac: u8,
}

impl CycleEstimation {
    fn base(base: u8) -> CycleEstimation {
        CycleEstimation {
            eac: 0,
            base,
            transfers: 0,
        }
    }
    fn with_transfers(cycles: u8, transfers: u8) -> CycleEstimation {
        CycleEstimation {
            eac: 0,
            base: cycles,
            transfers: transfers,
        }
    }
    fn with_eac(base_cycles: u8, transfers: u8, operand: &Operand) -> CycleEstimation {
        CycleEstimation {
            base: base_cycles,
            eac: estimate_eac(operand),
            transfers: transfers,
        }
    }
}

pub fn estimate_cycles_of(instr: &Instruction) -> CycleEstimation {
    match instr.op_code {
        OpCode::Jb
        | OpCode::Je
        | OpCode::Jl
        | OpCode::Jo
        | OpCode::Jp
        | OpCode::Js
        | OpCode::Jbe
        | OpCode::Jle
        | OpCode::Jnb
        | OpCode::Jne
        | OpCode::Jnl
        | OpCode::Jno
        | OpCode::Jnp
        | OpCode::Jns
        | OpCode::Jcxz
        | OpCode::Jnbe
        | OpCode::Jnle
        | OpCode::Loop
        | OpCode::Loopz
        | OpCode::Loopnz => panic!("TODO unknown estimation"),
        OpCode::Add | OpCode::Sub
            if instr.left_operand.is_reg() && instr.right_operand.is_reg() =>
        {
            CycleEstimation::base(3)
        }
        OpCode::Add | OpCode::Sub
            if instr.left_operand.is_reg() && instr.right_operand.is_eac() =>
        {
            CycleEstimation::with_eac(9, 1, &instr.right_operand)
        }
        OpCode::Add | OpCode::Sub
            if instr.left_operand.is_eac() && instr.right_operand.is_reg() =>
        {
            CycleEstimation::with_eac(16, 2, &instr.left_operand)
        }
        OpCode::Add | OpCode::Sub
            if instr.left_operand.is_acc_reg() && instr.right_operand.is_immediate() =>
        {
            CycleEstimation::base(4)
        }
        OpCode::Add | OpCode::Sub
            if instr.left_operand.is_reg() && instr.right_operand.is_immediate() =>
        {
            CycleEstimation::base(4)
        }
        OpCode::Add | OpCode::Sub
            if instr.left_operand.is_eac() && instr.right_operand.is_immediate() =>
        {
            CycleEstimation::with_eac(17, 2, &instr.left_operand)
        }
        OpCode::Add | OpCode::Sub => panic!("invalid instruction {:?}", instr),
        OpCode::Cmp if instr.left_operand.is_reg() && instr.right_operand.is_reg() => {
            CycleEstimation::base(3)
        }
        OpCode::Cmp if instr.left_operand.is_reg() && instr.right_operand.is_eac() => {
            CycleEstimation::with_eac(9, 1, &instr.right_operand)
        }
        OpCode::Cmp if instr.left_operand.is_eac() && instr.right_operand.is_reg() => {
            CycleEstimation::with_eac(9, 1, &instr.left_operand)
        }
        // acc is also here
        OpCode::Cmp if instr.left_operand.is_reg() && instr.right_operand.is_immediate() => {
            CycleEstimation::base(4)
        }
        OpCode::Cmp if instr.left_operand.is_eac() && instr.right_operand.is_immediate() => {
            CycleEstimation::with_eac(10, 1, &instr.left_operand)
        }
        OpCode::Cmp => panic!("invalid cmp instruction {:?}", instr),
        OpCode::Mov if instr.left_operand.is_mem_ref() && instr.right_operand.is_acc_reg() => {
            CycleEstimation::with_transfers(10, 1)
        }
        OpCode::Mov if instr.left_operand.is_acc_reg() && instr.right_operand.is_mem_ref() => {
            CycleEstimation::with_transfers(10, 1)
        }

        OpCode::Mov if instr.left_operand.is_reg() && instr.right_operand.is_reg() => {
            CycleEstimation::base(2)
        }
        OpCode::Mov if instr.left_operand.is_reg() && instr.right_operand.is_eac() => {
            CycleEstimation::with_eac(8, 1, &instr.right_operand)
        }
        OpCode::Mov if instr.left_operand.is_eac() && instr.right_operand.is_reg() => {
            CycleEstimation::with_eac(9, 1, &instr.left_operand)
        }
        OpCode::Mov if instr.left_operand.is_reg() && instr.right_operand.is_immediate() => {
            CycleEstimation::base(4)
        }
        OpCode::Mov if instr.left_operand.is_eac() && instr.right_operand.is_immediate() => {
            CycleEstimation::with_eac(10, 1, &instr.left_operand)
        }
        OpCode::Mov => panic!("unknown mov instr {:?}", instr),
    }
}
