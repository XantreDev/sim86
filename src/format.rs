use super::structs::*;

pub trait Formattable {
    fn format(&self) -> String;
}

impl Formattable for Register {
    fn format(&self) -> String {
        match self {
            Register::Ax => "ax",
            Register::Ah => "ah",
            Register::Al => "al",

            Register::Bx => "bx",
            Register::Bh => "bh",
            Register::Bl => "bl",
            Register::Bp => "bp",

            Register::Cx => "cx",
            Register::Cl => "cl",
            Register::Ch => "ch",
            Register::Cs => "cs",

            Register::Dx => "dx",
            Register::Dl => "dl",
            Register::Dh => "dh",
            Register::Di => "di",

            Register::Ip => "ip",
            Register::Si => "si",
            Register::Sp => "sp",
        }
        .to_string()
    }
}
impl Formattable for OpCode {
    fn format(&self) -> String {
        match self {
            OpCode::Add => "add",
            OpCode::Cmp => "cmp",
            OpCode::Mov => "mov",
            OpCode::Sub => "sub",

            OpCode::Je => "je",
            OpCode::Jl => "jl",
            OpCode::Jle => "jle",
            OpCode::Jb => "jb",
            OpCode::Jbe => "jbe",
            OpCode::Jp => "jp",
            OpCode::Jo => "jo",
            OpCode::Js => "js",
            OpCode::Jne => "jne",
            OpCode::Jnl => "jnl",
            OpCode::Jnle => "jnle",
            OpCode::Jnb => "jnb",
            OpCode::Jnbe => "jnbe",
            OpCode::Jnp => "jnp",
            OpCode::Jno => "jno",
            OpCode::Jns => "jns",
            OpCode::Loop => "loop",
            OpCode::Loopz => "loopz",
            OpCode::Loopnz => "loopnz",
            OpCode::Jcxz => "jcxz",
        }
        .to_string()
    }
}

impl Formattable for Operand {
    fn format(&self) -> String {
        fn pad_displacement(immediate: &Option<i16>) -> String {
            match immediate {
                Some(imm) if *imm > 0 => format!("+{}", imm),
                Some(imm) if *imm < 0 => imm.to_string(),
                _ => "".to_string(),
            }
        }
        match self {
            Self::Register(reg) => reg.format(),

            Self::Address(Some(a), Option::None, Option::None) => format!("[{}]", a.format()),
            Self::Address(Option::None, Option::None, Some(displacement)) => {
                format!("[{}]", displacement)
            }
            Self::Address(Some(a), Option::None, immediate) => {
                format!("[{}{}]", a.format(), pad_displacement(immediate))
            }
            Self::Address(Some(a), Some(b), immediate) => {
                format!(
                    "[{}+{}{}]",
                    a.format(),
                    b.format(),
                    pad_displacement(immediate)
                )
            }
            Self::Address(_, _, _) => panic!("unsupported {:?}", self),

            Self::Empty => "".to_string(),
            Self::JumpDisplacement(displacement) => {
                if displacement + 2 > 0 {
                    format!("+{}+0", displacement + 2)
                } else if displacement + 2 == 0 {
                    "+0".to_string()
                } else {
                    format!("{}+0", displacement + 2)
                }
            }
            Self::Immediate(immediate) => immediate.to_string(),
        }
    }
}

impl Formattable for Instruction {
    fn format(&self) -> String {
        if self.op_code.is_jump() {
            assert!(matches!(self.right_operand, Operand::Empty));
            assert!(matches!(self.right_operand, Operand::JumpDisplacement(_)));
            return format!("{} {}", self.op_code.format(), self.left_operand.format());
        }

        format!(
            "{} {}, {}",
            self.op_code.format(),
            self.left_operand.format(),
            self.right_operand.format()
        )
    }
}
