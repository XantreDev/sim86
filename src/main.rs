use std::{fs, io::Write, path::Path};

mod decoder;
mod estimation;
mod format;
mod simulator;
mod structs;

use structs::*;

use format::Formattable;

use crate::estimation::Architecture;

fn process_binary<T: Iterator<Item = u8>>(mut content_iter: T) -> Vec<Instruction> {
    use decoder;

    let mut result: Vec<Instruction> = Vec::with_capacity(2048);
    loop {
        let Some(instr) = decoder::decode_instruction(&mut content_iter) else {
            return result;
        };
        result.push(instr);
    }
}

fn write_instructions<T: Write>(instructions: Vec<Instruction>, out: &mut T) {
    writeln!(out, "bits 16\n").unwrap();
    instructions.iter().for_each(|it| {
        out.write((it.format() + "\n").as_bytes())
            .expect("must write");
    });
}

fn read_file<T: Into<String>>(file_path: T) -> Vec<u8> {
    let Ok(content) = std::fs::read(file_path.into()) else {
        panic!("Failed to read file");
    };

    content
}

struct DumpMemoryParams {
    path: String,
}
struct EstimateCyclesMode {
    mode: Architecture,
}

enum CliMode {
    Disasm,
    Exec(Option<DumpMemoryParams>, Option<EstimateCyclesMode>),
}

enum CliOptionState<T: Sized> {
    Unseen,
    SeenBefore,
    Parsed(T),
}

impl<T: Sized> CliOptionState<T> {
    fn is_seen_before(self: &Self) -> bool {
        match self {
            Self::SeenBefore => true,
            _ => false,
        }
    }
}

impl<T: Sized> Default for CliOptionState<T> {
    fn default() -> Self {
        CliOptionState::Unseen
    }
}

fn main() {
    let args = std::env::args().skip(1).collect::<Vec<String>>();
    let mut positional_args = args.iter().filter(|it| !it.starts_with("--"));
    let mode = match positional_args.next() {
        Some(str) if str == "disasm" => CliMode::Disasm,
        Some(str) if str == "exec" => {
            let mut iter = args.iter();

            let mut dump_cli_flag: CliOptionState<DumpMemoryParams> = CliOptionState::default();

            let mut cycles_estimation_option: CliOptionState<EstimateCyclesMode> =
                CliOptionState::default();

            loop {
                let Some(value) = iter.next() else {
                    break;
                };

                if matches!(dump_cli_flag, CliOptionState::Unseen) && value == "--dump" {
                    dump_cli_flag = CliOptionState::SeenBefore
                } else if matches!(dump_cli_flag, CliOptionState::SeenBefore) {
                    dump_cli_flag = CliOptionState::Parsed(DumpMemoryParams {
                        path: value.clone(),
                    });
                } else if matches!(cycles_estimation_option, CliOptionState::Unseen)
                    && value == "--estimate-cycles"
                {
                    cycles_estimation_option = CliOptionState::SeenBefore;
                } else if matches!(cycles_estimation_option, CliOptionState::SeenBefore)
                    && value == "8086"
                {
                    cycles_estimation_option = CliOptionState::Parsed(EstimateCyclesMode {
                        mode: Architecture::X8086,
                    });
                } else if matches!(cycles_estimation_option, CliOptionState::SeenBefore)
                    && value == "8088"
                {
                    cycles_estimation_option = CliOptionState::Parsed(EstimateCyclesMode {
                        mode: Architecture::X8088,
                    })
                } else if matches!(cycles_estimation_option, CliOptionState::SeenBefore) {
                    panic!("invalid option for cycle estimation: {}", value)
                }
            }
            let dump_params = match dump_cli_flag {
                CliOptionState::Unseen => None,
                CliOptionState::SeenBefore => panic!("no path for dump argument"),
                CliOptionState::Parsed(params) => Some(params),
            };
            let cycles_estimation_param = match cycles_estimation_option {
                CliOptionState::Unseen => None,
                CliOptionState::SeenBefore => {
                    panic!("no value for cycles estimation option (expected '8086' or '8088')")
                }
                CliOptionState::Parsed(params) => Some(params),
            };
            CliMode::Exec(dump_params, cycles_estimation_param)
        }
        Some(str) => {
            std::println!("Unknown mode: {}", str);
            return;
        }
        _ => {
            std::println!("Please at least two args");
            return;
        }
    };
    let Some(path) = positional_args.last() else {
        std::println!("Please provide assembly file path");
        return;
    };

    let content = read_file(path);
    #[cfg(target_endian = "big")]
    println!("This is a BigEndian system");

    #[cfg(target_endian = "little")]
    println!("This is a Little endian system");

    let mut stdout = &mut std::io::stdout().lock();
    match mode {
        CliMode::Disasm => {
            let content_iter = content.iter().map(|it| *it);
            let instrucitons = process_binary(content_iter);
            write_instructions(instrucitons, &mut stdout);
        }
        CliMode::Exec(dump_params, estimation) => {
            use crate::simulator::{execute, SimulatorConfig};
            let memory = execute(
                content,
                Some(stdout),
                SimulatorConfig {
                    cycle_estimation_mode: estimation.map(|it| it.mode),
                },
            );
            match dump_params {
                Some(DumpMemoryParams { path }) => {
                    let mut file = fs::File::create(Path::new(&path)).expect("should open file");
                    let dump = memory.dump();
                    file.write_all(dump).expect("should write succesfully");
                }
                None => {}
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use core::str;
    use std::{
        fs,
        io::Write,
        path::Path,
        process::{Command, ExitStatus},
    };

    use crate::{estimation::Architecture, process_binary, read_file, write_instructions};

    fn compile_asm_if_not(path: &str, force_recompile: bool) {
        let is_exist = Path::new(path).exists();
        if is_exist && force_recompile {
            fs::remove_file(Path::new(path)).expect("must delete");
        };
        if force_recompile || !is_exist {
            let asm_path = format!("{}.asm", path);
            assert!(Path::new(&asm_path).exists());

            let execution_result = Command::new("nasm").args([asm_path]).output().unwrap();
            if !execution_result.status.success() {
                let help = str::from_utf8(execution_result.stderr.as_slice())
                    .unwrap_or("<failed to parse stdout>");
                panic!("failed to compile asm:\n {}", help);
            }
            assert!(execution_result.status.success());
            assert!(Path::new(&path).exists());
        }
    }

    #[test]
    fn test_parser() {
        let files = [
            "./input/listing_0037_single_register_mov",
            "./input/listing_0038_many_register_mov",
            "./input/listing_0039_more_movs",
            "./input/listing_0040_challenge_movs",
            "./input/listing_0041_add_sub_cmp_jnz",
            "./input/listing_0043_immediate_movs",
            "./input/listing_0044_register_movs",
            "./input/listing_0044_register_movs_extended",
        ];
        let force_recompilation = std::env::var("RECOMPILE").map_or(false, |it| it == "true");

        for file_path in files {
            compile_asm_if_not(file_path, force_recompilation);
            let file = read_file(file_path);
            let mut res: Vec<u8> = Vec::new();
            println!("checking '{}'", file_path);

            write_instructions(process_binary(file.iter().map(|it| *it)), &mut res);

            if !Path::new("./.test").exists() {
                fs::create_dir("./.test").unwrap();
            }

            fs::File::create("./.test/test.asm")
                .unwrap()
                .write_all(res.as_slice())
                .unwrap();

            Command::new("nasm")
                .args(["./.test/test.asm"])
                .output()
                .unwrap();

            let nasm_file = read_file("./.test/test".to_string());
            assert!(nasm_file == file);
        }
    }

    #[test]
    fn test_execution() {
        let files = [
            "./input/listing_0043_immediate_movs",
            "./input/listing_0044_register_movs",
            "./input/listing_0044_register_movs_extended",
            "./input/listing_0046_add_sub_cmp",
            "./input/listing_0047_challenge_flags",
            "./input/listing_0048_ip_register",
            "./input/listing_0049_conditional_jumps",
            "./input/listing_0050_challenge_jumps",
            "./input/listing_0051_memory_mov",
            "./input/listing_0052_memory_add_loop",
            "./input/listing_0053_add_loop_challenge",
            "./input/listing_0054_draw_rectangle",
            "./input/listing_0055_challenge_rectangle",
        ];
        let force_recompilation = std::env::var("RECOMPILE").map_or(false, |it| it == "true");

        for file_path in files {
            compile_asm_if_not(file_path, force_recompilation);
            let file = read_file(file_path);
            let mut res: Vec<u8> = Vec::new();
            println!("evaluating '{}'", file_path);

            use crate::simulator::{execute, SimulatorConfig};
            execute(file, Some(&mut res), SimulatorConfig::default());
            insta::assert_snapshot!(str::from_utf8(res.as_slice()).unwrap());
        }
    }

    #[test]
    fn test_estimation() {
        let files = [
            "./input/listing_0056_estimating_cycles",
            "./input/listing_0057_challenge_cycles",
        ];
        let force_recompilation = std::env::var("RECOMPILE").map_or(false, |it| it == "true");

        for file_path in files {
            compile_asm_if_not(file_path, force_recompilation);
            let file = read_file(file_path);
            let mut res = Vec::new();

            write!(res, "X8086\n\n").unwrap();
            use crate::simulator::{execute, SimulatorConfig};
            execute(
                file.clone(),
                Some(&mut res),
                SimulatorConfig {
                    cycle_estimation_mode: Some(Architecture::X8086),
                },
            );

            write!(res, "\n\nX8088\n\n").unwrap();

            execute(
                file,
                Some(&mut res),
                SimulatorConfig {
                    cycle_estimation_mode: Some(Architecture::X8088),
                },
            );

            insta::assert_snapshot!(str::from_utf8(res.as_slice()).unwrap());
        }
    }
}
