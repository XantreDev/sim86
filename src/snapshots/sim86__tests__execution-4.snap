---
source: src/main.rs
expression: "str::from_utf8(res.as_slice()).unwrap()"
---
mov bx, -4093 ; bx 0x0000 -> 0xf003
mov cx, 3841 ; cx 0x0000 -> 0x0f01
sub bx, cx ; bx 0xf003 -> 0xe102 | flags: None -> S
mov sp, 998 ; sp 0x0000 -> 0x03e6
mov bp, 999 ; bp 0x0000 -> 0x03e7
cmp bp, sp ; flags: S -> None
add bp, 1027 ; bp 0x03e7 -> 0x07ea | flags: None -> None
sub bp, 2026 ; bp 0x07ea -> 0x0000 | flags: None -> PZ

Final registers: 
ax 0x0000 (0)
bx 0xe102 (57602)
cx 0x0f01 (3841)
dx 0x0000 (0)
sp 0x03e6 (998)
bp 0x0000 (0)
si 0x0000 (0)
di 0x0000 (0)
ip: 0x0018 (24)
flags: PZ
